rust/lance-index/src/scalar.rs

```diff
@@ -477,13 +486,92 @@ impl AnyQuery for LabelListQuery {
+#[derive(Debug, Clone, PartialEq)]
+pub enum TextQuery {
+    /// Retrieve all row ids where the text contains the given string
+    StringContains(String),
+    // TODO: In the future we should be able to do string-insensitive contains
+    // as well as partial matches (e.g. LIKE 'foo%') and potentially even
+    // some regular expressions
+}
+
+impl AnyQuery for TextQuery {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn format(&self, col: &str) -> String {
+        format!("{}", self.to_expr(col.to_string()))
+    }
+
+    fn to_expr(&self, col: String) -> Expr {
+        match self {
+            Self::StringContains(substr) => Expr::ScalarFunction(ScalarFunction {
+                func: Arc::new(ContainsFunc::new().into()),
+                args: vec![
+                    Expr::Column(Column::new_unqualified(col)),
+                    Expr::Literal(ScalarValue::Utf8(Some(substr.clone()))),
+                ],
+            }),
+        }
+    }
+
+    fn dyn_eq(&self, other: &dyn AnyQuery) -> bool {
+        match other.as_any().downcast_ref::<Self>() {
+            Some(o) => self == o,
+            None => false,
+        }
+    }
+
+    fn needs_recheck(&self) -> bool {
+        true
+    }
+}
+
+/// The result of a search operation against a scalar index
+#[derive(Debug)]
+pub enum SearchResult {
+    /// The exact row ids that satisfy the query
+    Exact(RowIdTreeMap),
+    /// Any row id satisfying the query will be in this set but not every
+    /// row id in this set will satisfy the query, a further recheck step
+    /// is needed
+    AtMost(RowIdTreeMap),
+    /// All of the given row ids satisfy the query but there may be more
+    ///
+    /// No scalar index actually returns this today but it can arise from
+    /// boolean operations (e.g. NOT(AtMost(x)) == AtLeast(NOT(x)))
+    AtLeast(RowIdTreeMap),
+}
+
+impl SearchResult {
+    pub fn row_ids(&self) -> &RowIdTreeMap {
+        match self {
+            Self::Exact(row_ids) => row_ids,
+            Self::AtMost(row_ids) => row_ids,
+            Self::AtLeast(row_ids) => row_ids,
+        }
+    }
+
+    pub fn is_exact(&self) -> bool {
+        matches!(self, Self::Exact(_))
+    }
+}
+
 /// A trait for a scalar index, a structure that can determine row ids that satisfy scalar queries
 #[async_trait]
 pub trait ScalarIndex: Send + Sync + std::fmt::Debug + Index + DeepSizeOf {
     /// Search the scalar index
     ///
     /// Returns all row ids that satisfy the query, these row ids are not necessarily ordered
-    async fn search(&self, query: &dyn AnyQuery) -> Result<RowIdTreeMap>;
+    async fn search(&self, query: &dyn AnyQuery) -> Result<SearchResult>;
+
+    /// Returns true if the query can be answered exactly
+    ///
+    /// If false is returned then the query still may be answered exactly but if true is returned
+    /// then the query must be answered exactly
+    fn can_answer_exact(&self, query: &dyn AnyQuery) -> bool;
```

Comment:
Do we actually need this with the changes ScalarIndexExpr?

Reply:
No.  Right now we rely on `AnyQuery::needs_recheck`.  I don't love that pattern though.  `AnyQuery` isn't really the right spot to make that check.  It works today because there is only one index that can answer `TextQuery` but there could (in theory) be indices in the future that can answer `TextQuery` exactly.  Or maybe a more likely scenario is an index that can answer `SargableQuery` inexactly.  I'd like to eventually get rid of `AnyQuery::needs_recheck` but I couldn't quite work it out to use `ScalarIndex::can_answer_exact` because today we need to make the decision early in the planning process before we have access to the scalar index.  So I'm leaving this in as the long term goal.